contracts/Staking.sol

```diff
@@ -0,0 +1,256 @@
+     */
+    uint256 public apr;
+
+    /**
+     * @dev contain how many reward tokens accumulated
+     */
+    uint256 public rewardPerTokenStored;
+
+    /**
+     * @dev contain how many tokens stake
+     * at the monment
+     */
+    uint256 public stakingTotalAmount;
+
+    IERC20 public token;
+
+    /**
+     * @dev Set 'token' IERC20 to interact with thrid party token
+     *
+     * @param token_ - of ERC20 contract
+     * @notice set staking params
+     */
+    constructor(address token_) {
+        require(
+            token_.isContract(),
+            "Error : Incorrect address , only contract address"
+        );
+        token = IERC20(token_);
+        stakingParams.stakingPeriod = 365 days;
+        stakingParams.feePercent = 40;
+        stakingParams.cooldownPeriod = 10 days;
+    }
+
+    /**
+     * @dev run staking period and transfer max reward
+     * value to this contract
+     *
+     * @param start_ - input time when staking period
+     * will be start
+     * @param rewardsAmount_ - max reward on this period
+     * @param apr_ - how many tokens wiil be collect
+     * per second
+     *
+     * @notice function can call only owner of SC
+     */
+    function setRewards(
+        uint256 start_,
+        uint256 rewardsAmount_,
+        uint256 apr_
+    ) external override onlyOwner {
+        require(
+            rewardsAmount_ <= MAX_REWARD,
+            "Error : reward above limit - 500000"
+        );
+        require(
+            start_ >= block.timestamp,
+            "Error : start time must be greater than current time"
+        );
+        require(
+            rewardsAmount_ > 0 && start_ > 0 && apr_ > 0,
+            "Error : one of params equal to 0"
+        );
+        stakingParams.stakingStartDate = start_;
+        stakingParams.stakingFinishDate =
+            stakingParams.stakingStartDate +
+            stakingParams.stakingPeriod;
+        apr = apr_;
+        token.safeTransferFrom(msg.sender, address(this), rewardsAmount_);
+        emit SetRewards(start_, rewardsAmount_, apr_);
+    }
+
+    /**
+     * @dev function allow user stake tokens
+     *
+     * @param amount_ - how many tokens user want to stake
+     *
+     * @notice function can call by any one , but sum of
+     * all staking tokens can't be higher than 'MAX_STAKING_POOL'
+     */
+    function stake(uint256 amount_) external override {
+        require(
+            block.timestamp >
+                investorList[msg.sender].startDate +
+                    stakingParams.cooldownPeriod ||
+                investorList[msg.sender].startDate < 0,
+            "Error : for re-staking wait 10 days"
+        );
+        require(amount_ > 0, "Error : you can't stake 0 tokens");
+        require(
+            stakingParams.stakingStartDate > 0,
+            "Error : staking has not started yet"
+        );
+        require(
+            stakingParams.stakingStartDate + stakingParams.stakingPeriod >
```

Comment:
you have stakingFinishDate variable in stakingParams, just use stakingFinishDate variable